File: interview/leetcode/day03042026/LetterCombination.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        phone = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        res = list()

        def backtrack(digits: str, digitIdx: int, phone: dict, currComb: str):
            if len(currComb) == len(digits):
                res.append(currComb)
                return

            currLetter = digits[digitIdx]

            for c in phone[currLetter]:
                backtrack(digits, digitIdx+1, phone,
                          currComb + c)

        backtrack(digits, 0, phone, "")
        return res


sol = Solution()
print(sol.letterCombinations("23"))

File: interview/leetcode/day19012026/LetterCombination.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class LetterCombination {
  private static final String[] KEYPADS = {
      "", // 0,
      "", // 1
      "abc", // 2
      "def", // 3
      "ghi", // 4
      "jkl", // 5
      "mno", // 6
      "pqrs", // 7
      "tuv", // 8
      "wxyz", // 9
  };

  public List<String> letterCombinations(String digits) {
    List<String> result = new ArrayList<>();
    backtrack(result, digits, new ArrayList<>(), 0);
    return result;
  }

  public void backtrack(List<String> result, String digits, List<Character> temp, int idx) {
    if (idx == digits.length()) {
      result.add(temp.stream().map(String::valueOf).collect(Collectors.joining()));
      return;
    }

    String letters = KEYPADS[digits.charAt(idx) - '0'];
    for (char letter : letters.toCharArray()) {
      temp.add(letter);
      backtrack(result, digits, temp, idx + 1);
      temp.remove(temp.size() - 1);
    }
  }

  public static void main(String[] args) {
    LetterCombination letterCombination = new LetterCombination();
    String digits = "23";
    List<String> result = letterCombination.letterCombinations(digits);
    System.out.println(result);
  }
}
